Make Ruby 2.4.10 work

This Ruby version doesn't work locally. To re-create:

    docker build . --build-arg RUBY_VERSION=2.4.10 

I think the error is something to do with `uname -P` returning `unknown` in the 2.4.10 image and that preventing the nokogiri build from completing, from `/usr/local/bundle/gems/nokogiri-1.6.6.4/ext/nokogiri/tmp/aarch64-unknown-linux-gnu/ports/libxslt/1.1.28/configure.log`:

```
configure: WARNING: if you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used
checking build system type... ./config.guess: unable to guess system type

This script, last modified 2007-05-17, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://savannah.gnu.org/cgi-bin/viewcvs/*checkout*/config/config/config.guess
and
  http://savannah.gnu.org/cgi-bin/viewcvs/*checkout*/config/config/config.sub

If the version you run (./config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2007-05-17

uname -m = aarch64
uname -r = 6.4.16-linuxkit
uname -s = Linux
uname -v = #1 SMP PREEMPT Wed Oct 25 16:32:24 UTC 2023

/usr/bin/uname -p = 
/bin/uname -X     = 

hostinfo               = 
/bin/universe          = 
/usr/bin/arch -k       = 
/bin/arch              = 
/usr/bin/oslevel       = 
/usr/convex/getsysinfo = 

UNAME_MACHINE = aarch64
UNAME_RELEASE = 6.4.16-linuxkit
UNAME_SYSTEM  = Linux
UNAME_VERSION = #1 SMP PREEMPT Wed Oct 25 16:32:24 UTC 2023
configure: error: cannot guess build type; you must specify one
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -16,6 +16,12 @@ RUN if ruby -e 'exit RUBY_VERSION < "2.6"'; then \
     else \
         gem install bundler -v "$(tail -n1 Gemfile.lock | tr -d ' ')"; \
     fi
+
+# I can't get nokogiri to build on the older versions with the build-in libraries.
+RUN if ruby -e 'exit RUBY_VERSION < "2.5"'; then \
+        bundle config build.nokogiri "--use-system-libraries --with-xml2-include=/usr/local/opt/libxml2/include/libxml2"; \
+    fi
+
 RUN bundle
 
 CMD ["bundle", "exec", "rake"]

--- a/Rakefile
+++ b/Rakefile
@@ -14,12 +14,11 @@ def normalize_version(version)
 end
 
 # Helper method to run docker-compose and capture the output, including build logs
-def docker_compose_up(service_name)
+def docker_compose_up(service_name, verbose: false)
   # Directing build output to a file
   build_log_file = "docker-build-#{service_name}.log"
-  quiet = ENV["VERBOSE"] ? "" : " > /dev/null"
+  quiet = verbose ? "" : " > /dev/null"
   system("docker-compose up --build --exit-code-from #{service_name} #{service_name} 2>&1 | tee #{build_log_file}#{quiet}")
-  build_log_file # Returning the log file path for reference
 end
 
 # Helper method to extract test results from container logs
@@ -62,7 +61,7 @@ task "test:all" do
   ruby_versions.each do |version|
     threads << Thread.new do
       service_name = "ruby_#{normalize_version(version)}"
-      docker_compose_up(service_name)
+      docker_compose_up(service_name, verbose: ENV["VERBOSE"])
       output = docker_compose_results(service_name)
 
       if output.include?(", 0 failures, 0 errors")
@@ -85,21 +84,18 @@ task "test:one", [:version] do |t, args|
   raise ArgumentError, "You must specify a Ruby version." unless version
   service_name = "ruby_#{normalize_version(version)}"
 
-  # Run the container in detached mode
-  build_log_file = docker_compose_up(service_name)
+  log = docker_compose_up(service_name, verbose: ENV["VERBOSE"])
 
-  # Now we can use docker compose to wait for the service to finish
-  system("docker compose up --exit-code-from #{service_name}")
-  system("docker compose logs #{service_name}")
+  output = docker_compose_results(service_name)
 
-  if ENV["VERBOSE"]
-    puts "Docker build logs for #{service_name}:"
-    puts File.read(build_log_file)
+  if output.include?(", 0 failures, 0 errors")
+    puts "Ruby #{version}: ✅ passed"
+  else
+    puts "Ruby #{version} ❌ failed, run this version using `rake \"test:one[#{version}]\"`"
   end
 
   # Stop and remove the container
   docker_compose_down(service_name)
 end
-
 
 task :default => :test

--- a/oas_agent.gemspec
+++ b/oas_agent.gemspec
@@ -27,11 +27,15 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
 
-  spec.add_development_dependency "msgpack", "~> 1.6.0"
+  # Because we're asking clients to include this gem in their app with an
+  # unknown Ruby version we can't pin the version here as it could conflict with
+  # the environment the gem is runnign in.
+  spec.add_dependency "msgpack", "< 1.4.2"
 
+  # Development dependencies must be version specced to work from Ruby 1.9.3 up to Ruby head
   spec.add_development_dependency "rubocop"
   spec.add_development_dependency "bundler", "~> 2.0"
   spec.add_development_dependency "rake"
   spec.add_development_dependency "minitest", "~> 5.0"
-  spec.add_development_dependency "nokogiri", "~> 1.12.0"
+  spec.add_development_dependency "nokogiri"
 end

